### ISSUE TITLE
Disallowing two outputs with `named_pubs` possibly too restrictive

### ISSUE THREAD

**Issue by bucko13** [buidl-bitcoin/buidl-python#158]
There is a check on multisig PSBTs [here](https://github.com/buidl-bitcoin/buidl-python/blob/c0b7d573a6411341be0485606ea17c75ec631aaa/buidl/psbt.py#L868C1-L873C22) that raises an error if there are two  change addresses. Unfortunately, how this is checked is a little more open than that, where it's not checking for actual change addresses but rather two outputs to "named pubs", i.e. outputs to the source wallet. But unfortunately this causes a failure if you're trying to do something like a test transaction.

Consider the scenario: you want to check signing in your multisig wallet, so you get a new address for your own wallet. You've got 1 utxo of value 1btc, you send yourself .1 btc for fun and let the wallet figure out how much fees should be and subtract that from the remaining 0.9 to determine the change. Even though the two outputs would come from different descriptors, they are still from the same global xpubs/wallet and will result in an error in this check.

**Comment by mflaxman** [comment]
Couldn't you just not name the xpubs in the psbt on the .1 "spend"? Then that change/fee would calculate automatically and it would describe the transaction as you did ("You've got 1 utxo of value 1btc, you send yourself .1 btc for fun and let the wallet figure out how much fees should be and subtract that from the remaining 0.9 to determine the change.")

I'm not against multiple change outputs in principle, but in practice this code is highly complex as-is and I have trouble coming up with a real use-case. Making the code-path more complex is going to increase the odds that a bug is introduced, which is something that [has happened](https://benma.github.io/2021/02/09/coldcard-multisig-vulnerability.html) in other libraries.

**Comment by mflaxman** [comment]
> It's not really clear to me how this type of swap attack is relevant here. If it's just a matter of complexity leading to possible attack vectors, it could be said that this adds additional complexity as well and so I don't see much of a difference.

Read the code path, I spent sooooo much time writing and auditing this. Even jumping back in for the last hour it's very hard to reason about what a simple change would do. The logic is as simplified as I could make it at the time, but still far more complex than I'd like. Keeping it simple helps increase the odds it's working as designed.

It might be possible to just get rid of these lines, but it would require a big increase in test coverage:
https://github.com/buidl-bitcoin/buidl-python/blob/main/buidl/psbt.py#L796
https://github.com/buidl-bitcoin/buidl-python/blob/main/buidl/psbt.py#L868-L875

It's also worth noting that it's possible the transaction you're describing might still not validate even if you did make these changes. Have you tried? The change detection assumes the most stringent choices, IDK if what you're describing would otherwise get through.

>  if the workaround is simply to remove some optional information from the PSBT, which is something an attacker could do anyway, what is this protecting us from?

That information changes the human description of the PSBT, which is the main job of the helper method. You're still free to sign or alter any PSBT in buidl. The description of a PSBT without a named output will say that it's a spend (not validated change), the description of a PSBT with a named output must validate that it is in fact change. That's the most important thing this method does.

--

In doing this, I discovered that `bip32_derivs` is not being used in `_describe_basic_multisig_outputs()` (it's there from a refactor to match `_describe_basic_multisig_inputs()`). It'd be great to get rid of that code, if I'm 100% sure it's not relevant and that the change detection still works as intended.

**Comment by bucko13** [comment]
> That information changes the human description of the PSBT, which is the main job of the helper method. You're still free to sign or alter any PSBT in buidl.

Yeah, you're right (of course. you wrote it! 😅). This part wasn't clear to me as it seemed like a validation on signing the way the behavior operates (raising an exception rather than adding to a description)

Since this is actually just a description, maybe it would be better to aggregate warnings and include them in a final description output. It's possible the original implementor of the code that made this call wasn't expecting this to cause signing to fail because of this naming which resulted in unwanted behavior that broke signing. We had other places where signing was working with the same PSBT so it was confusing that I couldn't reproduce, which is how I realized that the only place in buidl where describe is called is in the multisig wallet cli app.

I would expect a `validate` method to fail and a `describe` method to provide outputs allowing consumer to decide how to proceed, but that's just me. Another advantage of aggregating errors is that you can get all of them at once and address them all at once. This is particularly useful for extremely secure, "frozen" storage setups where it's labor intensive to have to adjust PSBTs only to learn there are other errors to fix (this can be fixed operationally as well, but I think the API helps with identifying operational changes). 

The workaround of stripping out BIP32 derivation information for all but one output seems to work for now, so I'll close the issue. 

I still think it's worth considering how some of these helper methods and this library might be used in more complex setups and perhaps considering loosening up errors at least into warnings. I totally appreciate (literally, thank you!) the concern for safety and approach to avoiding footguns, but one of the scary and great things about multisig is all of the new use cases it opens up. The maximum footgun protections make most sense in the simple setup with one main quorum owner/maintainer, but when you start moving into areas like collaborative custody things get more complex. E.g. whereas you can't see a situation like the one described here, when you are just one of many signers and are not responsible for the creation of transactions in the first place, weird but still "desirable" things can happen. Add likely future scenarios like inputs contributed from multiple parties or outputs to multiple parties that include some of the signers who might want to verify their own change output amongst many, and things get murkier still. I think notifying about these "sketchy transaction" scenarios is definitely useful, but perhaps exceptions are overkill.

**Comment by mflaxman** [comment]
> Since this is actually just a description, maybe it would be better to aggregate warnings and include them in a final description output

I like the idea of aggregating warnings in theory, although in practice it's a little bit messy. For some of the errors/warnings, once a PSBT triggers one I can't say with 100% certainty that the checking on the rest remains accurate. That seems like danger zone territory. For now, you can always call the method, and have logic to catch the error.

One idea I had in how I built this was that if you have a change address problem, then you can always generate your PSBT transaction as a sweep so there is no change address. That logic is 10x simpler to reason about and you don't risk holding funds hostage to a validation bug. That's obviously not a good long-term solution, but it means that a very strict code-path has a get yourself out of trouble mode without having to write custom code and you can still have confidence that all the validation is working as intended.

Btw, for something similar to your use-case (sending to self), that is commonly done as-is by not giving the full path information for the self-spend (sparrow and others do this automatically if you just find a receive address and send some amount to it). So even though you have a list of xpubs, without claiming a spend is actually to oneself, the validation doesn't even know where to look (let alone verify) and just accepts it as spend (not change) and can display that to the user (which matches their original intention). IIRC different hardware wallets validate this transaction differently; I think I've seen some describe a transaction as just "TX fee with no spend" (IDK how they'd know without path info), which is confusing but most technically accurate.

--

All that said, I could still see a reason to do multiple changes (perhaps you're paying with a 10BTC UTXO and you want to send 1 BTC to a destination and then 9 changes to yourself of 1BTC each to break up a large UTXO). I don't think my safety checks about multiple changes are adding anything so I'm open to deleting those lines of code. I'm trying to think adversarially on this, and I guess one "attack" would be to make someone's spend destroy a bunch of BTC by creating a ton of dust change outputs. There's no benefit to the attacker, it's similar to a transaction paying a fee that is orders of magnitude too high (although that one has the benefit that if the attacker can eclipse you they can attempt to solo mine the transaction, however unlikely they may be to succeed). So in order to do this I'd have to introduce a new way to call this method that lists warnings and tells you if your TX is anything but 100% "safe." That's so complicated to explain to developers and end-users ("warning, your transaction has multiple change outputs. This could be used legitimately to break up a large UTXO, or maliciously to spread your change into unspendable dust. Proceed with caution or generate a new PSBT with <=1 change outputs to be safe.").
